Initial vector db implementation

snopoke opened the pull request:
This adds a vector DB implementation based off [langchain pgvector](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/pgvector.py) but backed by a custom Django model.

This can not be merged to main yet since our current production DB does not support the pgvector extension.

## Upgrade note
If you want to update without loosing all your existing data in postgres you can follow the instructions here: https://gist.github.com/snopoke/80e9a2e88ec493b7d4efae2578339f00

snopoke committed:
add pgvector requirement

snopoke committed:
initial embedding model

snopoke committed:
create pgvector extension

snopoke committed:
langchain VectorDb implementation

snopoke committed:
vectorstore tests and tweaks

snopoke committed:
remove comment

snopoke committed:
update pg image in CI

SmittieC commented on apps/vectordb/tests/test_vectorstore.py:
```diff
@@ -0,0 +1,143 @@
+        experiment=experiment,
+    )
+    output = docsearch.similarity_search("foo", k=1)
+    _remove_meta_fields(output)
```
Why's the meta fields removed here?

snopoke replied:
yes, just for testing so that we can compare the results without having to compare the metadata

SmittieC commented on apps/vectordb/tests/test_vectorstore.py:
```diff
@@ -0,0 +1,143 @@
+    assert output == [
+        (Document(page_content="foo", metadata={"page": "0"}), 0.0),
+        (Document(page_content="baz", metadata={"page": "2"}), 0.9290842232061864),
+        (Document(page_content="bar", metadata={"page": "1"}), 1.0246011368123038),
```
Interesting that this is larger than 1. I presume these scores refer to angles then?

snopoke replied:
I was a bit surprised by this too but then I checked and pgvector returns the cosine distance and not the cosine similarity (https://github.com/pgvector/pgvector/blob/d7354a86a8496a4432b9d27e2eaf5658274ed1f4/src/vector.c#L696)

### tl;dr

Similarity is the angle which ranges from -1 to 1
Distance is 1 - similarity which ranges from 0 to 2

See https://medium.com/@milana.shxanukova15/cosine-distance-and-cosine-similarity-a5da0e4d9ded for a good article on the difference.

SmittieC replied:
That makes sense, thanks!

snopoke committed:
Merge branch 'main' into sk/pg-vector

chidochipotle committed:
Add RAG minimum viable feature

chidochipotle committed:
Remove debug leftover print statements

chidochipotle committed:
Remove untested html support

chidochipotle committed:
Replacing files.all() by better suited files.exists()

chidochipotle committed:
Moving RAG embedding creation to celery task

chidochipotle committed:
Unfinished improvement:  Using prompt from the experiment
available_variables should be added as "source_material" using form_data , but I could not figure out how to od it with form_data.

chidochipotle committed:
Fixed several small issues commented by Simon in PR RAG MVP #374

snopoke committed:
Merge pull request #374 from chidochipotle/sk/pg-vector

RAG MVP

snopoke committed:
convert to using pymupdf4llm

snopoke committed:
use Django's file API to load file contents
